integrations/operator/controllers/resources/secretlookup/secretlookup.go

```diff
@@ -0,0 +1,113 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package secretlookup
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/gravitational/trace"
+	corev1 "k8s.io/api/core/v1"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	// secretScheme is the URI scheme for Kubernetes Secret Lookup
+	secretScheme = "secret"
+	secretPrefix = secretScheme + "://"
+
+	// AllowInclusionAnnotation is the annotation a secret must wear for the operator to allow looking up its content
+	// when reconciling a resource. Its value is either a comma-separated list of allowed resources, or a '*'.
+	AllowInclusionAnnotation = "resources.teleport.dev/allow-inclusion-from-cr"
+)
+
+// IsNeeded checks if a string starts with "secret://" and needs a secret lookup.
+func IsNeeded(value string) bool {
+	return strings.HasPrefix(value, secretPrefix)
+}
+
+// Try takes a URI such as "secret://secret-name/secret-key" and returns the data from the secret key.
+// To protect against someone abusing this to read from arbitrary secrets, the secret must be annotated with the
+// names of the CRs allowed to include it (or a wildcard).
+func Try(ctx context.Context, clt kclient.Client, name, namespace, uri string) (string, error) {
+	// Parse and check the URI
+	u, err := url.Parse(uri)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if u.Scheme != secretScheme {
+		return "", trace.BadParameter("invalid secret scheme %q", u.Scheme)
+	}
+
+	if u.Host == "" {
+		return "", trace.BadParameter("missing secret name")
+	}
+
+	if u.Path == "" {
+		return "", trace.BadParameter("missing secret key")
+	}
+	key := strings.TrimPrefix(u.Path, "/")
+
+	// Lookup the secret
+	secret := &corev1.Secret{}
+	if err := clt.Get(ctx, kclient.ObjectKey{Namespace: namespace, Name: u.Host}, secret); err != nil {
+		return "", trace.Wrap(err, "failed to lookup secret")
+	}
+
+	// Checking permission
+	if err := isInclusionAllowed(secret, name); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	secretData, ok := secret.Data[key]
+	if !ok {
+		return "", trace.BadParameter("secret %q is missing key %q", u.Host, key)
+	}
+
+	// Apparently we don't need to b64 decode ¯\_(ツ)_/¯
```

Comment:
Test seems to be failing on this.
```
--- FAIL: TestLookupSecret (0.00s)
    --- FAIL: TestLookupSecret/successful_lookup (0.00s)
        secretlookup_test.go:159: <nil>
        secretlookup_test.go:161: 
            	Error Trace:	/__w/teleport/teleport/integrations/operator/controllers/resources/secretlookup/secretlookup_test.go:161
            	Error:      	Not equal: 
            	            	expected: "test-key-value"
            	            	actual  : "dGVzdC1rZXktdmFsdWU="
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-test-key-value
            	            	+dGVzdC1rZXktdmFsdWU=
            	Test:       	TestLookupSecret/successful_lookup
```

Reply:
Thanks, the logic was good: the integration tests against a real Kubernetes API are working.

When I mocked the kubernetes client in the unit tests, I base64-encoded the content while it should have been the raw data.